bee-tangle/src/tangle.rs

```diff
@@ -0,0 +1,535 @@
+where
+    T: Clone,
+    H: Default,
+{
+    fn default() -> Self {
+        Self::new(H::default())
+    }
+}
+
+impl<T, H: Hooks<T>> Tangle<T, H>
+where
+    T: Clone,
+{
+    /// Creates a new Tangle.
+    pub fn new(hooks: H) -> Self {
+        Self {
+            vertices: TRwLock::new(HashMap::new()),
+
+            cache_queue: Mutex::new(LruCache::unbounded_with_hasher(DefaultHashBuilder::default())),
+            max_len: AtomicUsize::new(DEFAULT_CACHE_LEN),
+
+            hooks,
+        }
+    }
+
+    /// Create a new tangle with the given capacity.
+    pub fn with_capacity(self, cap: usize) -> Self {
+        Self {
+            cache_queue: Mutex::new(LruCache::with_hasher(cap + 1, DefaultHashBuilder::default())),
+            ..self
+        }
+    }
+
+    /// Change the maximum number of entries to store in the cache.
+    pub fn resize(&self, len: usize) {
+        self.max_len.store(len, Ordering::Relaxed);
+    }
+
+    /// Return a reference to the storage hooks used by this tangle.
+    pub fn hooks(&self) -> &H {
+        &self.hooks
+    }
+
+    async fn insert_inner(
+        &self,
+        message_id: MessageId,
+        message: Message,
+        metadata: T,
+        prevent_eviction: bool,
+    ) -> Option<MessageRef> {
+        let mut vertices = self.vertices.write().await;
+        let vtx = vertices.entry(message_id).or_insert_with(Vertex::empty);
+
+        if prevent_eviction {
+            vtx.prevent_eviction();
+        }
+
+        let msg = if vtx.message().is_some() {
+            None
+        } else {
+            let parents = message.parents().clone();
+
+            vtx.insert_message_and_metadata(message, metadata);
+            let msg = vtx.message().cloned();
+
+            let mut cache_queue = self.cache_queue.lock().await;
+
+            // Insert children for parents
+            for &parent in parents.iter() {
+                let children = vertices.entry(parent).or_insert_with(Vertex::empty);
+                children.add_child(message_id);
+
+                // Insert cache queue entry to track eviction priority
+                cache_queue.put(parent, ());
+            }
+
+            // Insert cache queue entry to track eviction priority
+            cache_queue.put(message_id, ());
+
+            msg
+        };
+
+        drop(vertices);
+
+        self.perform_eviction().await;
+
+        msg
+    }
+
+    /// Inserts a message, and returns a thread-safe reference to it in case it didn't already exist.
+    pub async fn insert(&self, message_id: MessageId, message: Message, metadata: T) -> Option<MessageRef> {
+        let exists = self.pull_message(&message_id, true).await;
+
+        let msg = self
+            .insert_inner(message_id, message.clone(), metadata.clone(), !exists)
+            .await;
+
+        self.vertices
+            .write()
+            .await
+            .get_mut(&message_id)
+            .expect("Just-inserted message is missing")
```

Comment:
could this happen if something runs between `insert_inner` and this?

Reply:
This is actually the race condition that I fixed last week by adding a flagging system that prevents eviction. So, no.